In [7]:
import pandas as pd, numpy as np, pathlib, os
from pathlib import Path

cur_path = Path()
os.chdir(cur_path)

#1)
def norm_data(data:list):
    'Takes in a list of data and normalizes it. In the event that the data max and min are the same it returns a ZeroDivisionError'
    data_min = min(data)
    data_max = max(data)
    if data_max-data_min == 0:
        print("Your dataset minimum and maximum are the same value, data can't be normalized as its all the same.")
        return ZeroDivisionError
    data = [(i-data_min)/(data_max-data_min) for i in data]
    return data

#2)
def stand_data(data:list):
    'Takes a list of data and standardizes it.'
    data_mean = np.mean(data)
    data_std_dev = np.std(data)
    data = [(i-data_mean)/data_std_dev for i in data]
    return data

ca_housing_data = pd.read_csv('calif_housing_data.csv')
display(ca_housing_data)


housing_median_age  total_bedrooms  households  median_income  \
0                      41           129.0         126         8.3252   
1                      21          1106.0        1138         8.3014   
2                      52           190.0         177         7.2574   
3                      52           235.0         219         5.6431   
4                      52           280.0         259         3.8462   
...                   ...             ...         ...            ...   
20635                  25           374.0         330         1.5603   
20636                  18           150.0         114         2.5568   
20637                  17           485.0         433         1.7000   
20638                  18           409.0         349         1.8672   
20639                  16           616.0         530         2.3886   

       median_house_value  
0                452600.0  
1                358500.0  
2                352100.0  
3                341300.0  
4                342200.0  
...                   ...  
20635             78100.0  
20636             77100.0  
20637             92300.0  
20638             84700.0  
20639             89400.0  

[20640 rows x 5 columns]

3A) This dataset has 20,640 rows.


3B) Since we are going to build a model that predicts the median house value our target vector is the median house value. As in we are trying to use the other variables to gain a deeper understanding about the patterns that result in what a house is worth. 

3C) As seen below, the feature we create is indicative of the average number of bedrooms per given household.

In [8]:
#3C)
ca_housing_data['avg_bedrooms'] = ca_housing_data['total_bedrooms']/ca_housing_data['households']
display(ca_housing_data)

housing_median_age  total_bedrooms  households  median_income  \
0                      41           129.0         126         8.3252   
1                      21          1106.0        1138         8.3014   
2                      52           190.0         177         7.2574   
3                      52           235.0         219         5.6431   
4                      52           280.0         259         3.8462   
...                   ...             ...         ...            ...   
20635                  25           374.0         330         1.5603   
20636                  18           150.0         114         2.5568   
20637                  17           485.0         433         1.7000   
20638                  18           409.0         349         1.8672   
20639                  16           616.0         530         2.3886   

       median_house_value  avg_bedrooms  
0                452600.0      1.023810  
1                358500.0      0.971880  
2                352100.0      1.073446  
3                341300.0      1.073059  
4                342200.0      1.081081  
...                   ...           ...  
20635             78100.0      1.133333  
20636             77100.0      1.315789  
20637             92300.0      1.120092  
20638             84700.0      1.171920  
20639             89400.0      1.162264  

[20640 rows x 6 columns]

In [16]:
#3D)
ca_housing_data_slice = ca_housing_data[['housing_median_age', 'median_income', 'avg_bedrooms']]
display(ca_housing_data_slice)

#3E)
ca_housing_data_slice = ca_housing_data_slice.apply(lambda x: stand_data(x))
display(ca_housing_data_slice)

housing_median_age  median_income  avg_bedrooms
0                      41         8.3252      1.023810
1                      21         8.3014      0.971880
2                      52         7.2574      1.073446
3                      52         5.6431      1.073059
4                      52         3.8462      1.081081
...                   ...            ...           ...
20635                  25         1.5603      1.133333
20636                  18         2.5568      1.315789
20637                  17         1.7000      1.120092
20638                  18         1.8672      1.171920
20639                  16         2.3886      1.162264

[20640 rows x 3 columns]

housing_median_age  median_income  avg_bedrooms
0                0.982143       2.344766     -0.153863
1               -0.607019       2.332238     -0.262936
2                1.856182       1.782699     -0.049604
3                1.856182       0.932968     -0.050417
4                1.856182      -0.012881     -0.033568
...                   ...            ...           ...
20635           -0.289187      -1.216128      0.076185
20636           -0.845393      -0.691593      0.459421
20637           -0.924851      -1.142593      0.048373
20638           -0.845393      -1.054583      0.157233
20639           -1.004309      -0.780129      0.136952

[20640 rows x 3 columns]